### ISSUE TITLE
Extmark can't render `\t` perfectly if line with highlights added by `matchadd` .

### ISSUE THREAD

**Issue by kevinhwang91** [neovim/neovim#20885]
### Neovim version (nvim -v)

NVIM v0.8.0

### Vim (not Nvim) behaves the same?

no

### Operating system/version

ArchLinux

### Terminal name/version

Alacritty

### $TERM environment variable

Alacritty

### Installation

repo

### How to reproduce the issue

`cat mini.lua`
```lua
	-- match1 with '\t'
        -- match2 with spaces
vim.fn.matchadd('Search', 'match')
local ns = vim.api.nvim_create_namespace('')
for i = 1, 2 do
    vim.api.nvim_buf_set_extmark(0, ns, i - 1, 0, {end_row = i, end_col = 0, hl_group = 'Visual'})
end
```
`nvim --clean +'so mini.lua' mini.lua`

### Expected behavior

![image](https://user-images.githubusercontent.com/17562139/198908148-1c229a2e-5173-4405-b285-a19d98bb680d.png)

### Actual behavior

![image](https://user-images.githubusercontent.com/17562139/198908205-b2565713-e8f3-4a88-8a0c-b139740b1587.png)

**Comment by zeertzjq** [comment]
Also applies to control characters.

**Comment by zeertzjq** [comment]
Fixed by #23742
